### ISSUE TITLE
Sync url option not working

### ISSUE THREAD

**Issue by Xachman** [eclipse-che/che#6514]
Where does the --url option go in the command? Im putting it after sync and it says "Error when trying to get workspace data for workspace named --url" I put it before sync and I get "ERROR: You passed an unknown command: --url".  Probably a better explanation of how to use these options  at http://www.eclipse.org/che/docs/ide/sync/index.html could help.

**Reproduction Steps:**
Run eclipse/che cli with docker with the sync command and the url option ex:
docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock -v <dataDir>:/data -v <syncDir>:/sync eclipse/che sync --url http://<ip>:<port> <workspaceID>

**OS and version:**
Windows 10    

**Diagnostics:** 
`WARN: Bound 'eclipse/che' to 'eclipse/che:5.18.0'                                                                       INFO: (che cli): 5.18.0 - using docker 17.06.2-ce / docker4windows                                                                                                                                                                              CLI:                                                                                                                     TTY:            true                                                                                                    Daemon:         /var/run/docker.sock                                                                                    Image:          eclipse/che                                                                                             Version:        5.18.0                                                                                                  Command:        info                                                                                                    Parameters:     info                                                                                                   Mounts:                                                                                                                  /data:          /c/projects/chedata                                                                                     /data/instance: not set                                                                                                 /data/backup:   not set                                                                                                 /repo:          not set                                                                                                 /assembly:      not set                                                                                                 /sync:          /c/projects/test-sync                                                                                   /unison:        not set                                                                                                 /chedir:        not set                                                                                                System:                                                                                                                  Docker:         docker4windows                                                                                          Proxy:          not set                                                                                                Internal:                                                                                                                CHE_VERSION:    5.18.0                                                                                                  CHE_HOST:       10.0.75.2                                                                                               CHE_INSTANCE:   /c/projects/chedata/instance                                                                            CHE_CONFIG:     /c/projects/chedata                                                                                     CHE_BACKUP:     /c/projects/chedata/backup                                                                              CHE_REGISTRY:   /version                                                                                                CHE_DEBUG:      false                                                                                                   IP Detection:   10.0.75.2                                                                                               Initialized:    false                                                                                                  Image Registry:                                                                                                          IMAGE_INIT=eclipse/che-init:5.18.0                                                                                      IMAGE_CHE=eclipse/che-server:5.18.0                                                                                     IMAGE_COMPOSE=docker/compose:1.8.1                                                                                      IMAGE_TRAEFIK=traefik:v1.3.0-rc3                                                                                        BOOTSTRAP_IMAGE_ALPINE=alpine:3.4                                                                                       BOOTSTRAP_IMAGE_CHEIP=eclipse/che-ip:5.18.0                                                                             UTILITY_IMAGE_CHEACTION=eclipse/che-action:5.18.0                                                                       UTILITY_IMAGE_CHEDIR=eclipse/che-dir:5.18.0                                                                             UTILITY_IMAGE_CHETEST=eclipse/che-test:5.18.0                                                                           UTILITY_IMAGE_CHEMOUNT=eclipse/che-mount:5.18.0                                                                        che.env: not initialized    `

**Comment by garagatyi** [comment]
The first argument of sync command is workspace name or workspace id.

**Comment by ghost** [comment]
From docs:

```
# Sync a workspace to local directory, <local-sync-point>
docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock
                    -v <path>:/data
                    -v <local-sync-point>:/sync
                    eclipse/che sync <workspace-name>
```
`--url` goes just after your workspace name. If not provided, some defaults are used.

**Comment by Xachman** [comment]
ok so when I add my url I get 

TypeError: Cannot read property 'getAgents' of undefined

ERROR: Error when trying to get workspace data for workspace named <workspace-name>

and then 

ERROR: (Eclipse Che action/list-workspaces): Unable to connect to the remote host 10.0.75.2 on port 8080. Please check the server is listening and that there is no network issue to reach this host. Full
error: Error: connect ECONNREFUSED 10.0.75.2:8080

my url option was 192.168.1.151:8900

So im totally confused about that last error looks like it may be the ip of the container with the port 8080

**Comment by Xachman** [comment]
my command output from docker inspect was 
`"Cmd": [
                "sync",
                "wksp-xc5z",
                "--url",
                "http://192.168.1.151:8900"
            ],
`

**Comment by ghost** [comment]
@Xachman if it's Docker for Windows, your IP should be 10.0.75.2. What IP do you use to access Che?

**Comment by Xachman** [comment]
Che is on another machine in my local network at 192.168.1.151 running on port 8900

Im trying to connect to it from docker for windows

**Comment by Xachman** [comment]
The command im running is (I hope this is what you mean by sync syntax):
`docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock -v /c/projects/chedata:/data -v /c/projects/test-sync:/sync eclipse/che sync wksp-xc5z --url http://192.168.1.151:8900 `             

if I change the url to be http://192.168.1.151:8901 ill get an error message:

Unable to connect to the remote host 192.168.1.151 on port 8901. Please check the server is listening and that there is no network issue to reach this host. Full error: Error: connect ECONNREFUSED 192.16
8.1.151:8901

I dont get this error when I use the correct url http://192.168.1.151:8900 I get the errors mentioned above

I can curl http://192.168.1.151:8900 and get the html output of the ide from inside a container in docker for windows.

My setup is a webserver using a local ip in my network.  Im not using any virtualization for the che part.

**Comment by ghost** [comment]
I don't see any reason why `--url http://192.168.1.151:8900` makes it connect to `10.x` IP.

Can you try deleting everything in `/c/projects/chedata` and give it another shot? I cannot reproduce the issue so far. If you are trying the wrong IP or URL, you can see `ECONNREFUSED`, right?

**Comment by Xachman** [comment]
If i try to connect to the wrong ip id get the ECONNREFUSED error.

Also just found this out.  If the workspace is not started it returns the error:

Error: Workspace should be in running state. Current state is STOPPED

If the workspace is starting I get the error:

Error: Workspace should be in running state. Current state is STARTING

If the workspace is started I get the errror:

TypeError: Cannot read property 'getAgents' of undefined

In each of these conditions I get 2 other errors:

ERROR: Error when trying to get workspace data for workspace named wksp-xc5z

and 

List of workspaces are:
ERROR: (Eclipse Che action/list-workspaces): Unable to connect to the remote host 10.0.75.2 on port 8080. Please check the server is listening and that there is no network issue to reach this host. Full
error: Error: connect ECONNREFUSED 10.0.75.2:8080

**Comment by ghost** [comment]
@Xachman yes, workspace should be running.

Can you record a short video showing the problem? I still cannot reproduce it. I have just started Che on my Ubuntu machine and then synced projects from a Windows laptop using 192.168.x.x:8080 as --url argument.

**Comment by Xachman** [comment]
My environment is the Che host is on an Ubuntu machine and im trying to sync a windows desktop.  That seems to be the reverse of what you said if you want to try that out.  Ill work on getting a video together.  I have been using power shell for the syncing but i also tried a bash shell and got the same results.

**Comment by ghost** [comment]
my setup is the same

**Comment by Xachman** [comment]
Want to reference this issue since this guy was getting the TypeError as well https://github.com/eclipse/che/issues/4975.  Ill try and get this video together this weekend.

**Comment by ghost** [comment]
In https://github.com/eclipse/che/issues/4975 we never got any additional info. 

I failed to reproduce it by running Che on Ubuntu and sync projects from windows. Both machines are in the same network. 

Does sync work for you on Ubuntu?

**Comment by Xachman** [comment]
Ok so I probably should have mentioned this before but im using a custom stack. I just tried the standard java stack and it worked fine.  So based on this and the other issue im going to say it has to do with the stack itself. Im not 100% sure what items need to be selected to make this work.

**Comment by ghost** [comment]
@Xachman you docker recipe please?

**Comment by Xachman** [comment]
@eivantsov  How do I get the recipe for a stack I made?  Should I just paste the raw json/configuration here?

**Comment by Xachman** [comment]
https://hub.docker.com/r/xachman/cakephp3-che/ Im working with this image now just building a stack based on it and im getting all the same errors.  I added some things I saw from the java stack to this image that I thought may be missing from my xachman/cakephp3 image.

**Comment by ghost** [comment]
@Xachman I'll check your recipe and get back to you. Most likely, ssh connection to the container fails. So, we need to find out why.

**Comment by Xachman** [comment]
I created a new stack using that image and for whatever reason the syncing started working.  Im not sure whats different but  I kept both around so i could compare them.  

That being said im getting an error now when I try and sync the vendor folder after running composer update on my cakephp3 project. The initial sync went well but when all the dependencies are loaded into the vendor folder I get an error:

ERROR: Fatal error occurred (1)

One option here would be to ignore it in the unision user profile but I have no idea how to include my default.prf file.  
I had a look at this issue https://github.com/eclipse/che/issues/5658 but I have not been able to get it recognized.
